backends/ldap.go

```diff
@@ -0,0 +1,343 @@
+	}
+
+	userDN := searchResult.Entries[0].DN
+
+	userLdapClient, err := l.LDAPClientFactory(l)
+	if err != nil {
+		log.Debugf("LDAP user connection error: %s", err)
+		return false, err
+	}
+	defer func(ldapClient LDAPClient) {
+		err := ldapClient.Close()
+		if err != nil {
+			log.Errorf("LDAP user cleanup error: %s", err)
+		}
+	}(userLdapClient)
+
+	err = userLdapClient.Bind(userDN, password)
+	if err != nil {
+		log.Debugf("LDAP user bind error: %s", err)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (l LDAP) GetSuperuser(username string) (bool, error) {
+
+	//If there's no superuser filter, assume all privileges for all users.
+	if l.SuperuserFilter == "" {
+		return false, nil
+	}
+
+	searchRequest := ldap.NewSearchRequest(
+		l.BaseDN,
+		ldap.ScopeWholeSubtree,
+		ldap.NeverDerefAliases,
+		0,
+		0,
+		false,
+		fmt.Sprintf(l.SuperuserFilter, ldap.EscapeFilter(username)),
+		[]string{"dn"},
+		nil,
+	)
+
+	searchResult, err := l.LDAPClient.Search(searchRequest)
+	if err != nil {
+		log.Debugf("LDAP superuser search error: %s", err)
+		return false, nil
+	}
+	if len(searchResult.Entries) != 1 {
+		log.Debugf("LDAP superuser search returned %d entries", len(searchResult.Entries))
+		return false, err
+	}
+
+	return true, nil
+}
+
+// checks whether an MQTT topic matches a pattern with wildcards (+, #) according to MQTT spec rules.
+func matchMQTT(topic, pattern string) bool {
+	topicLevels := strings.Split(topic, "/")
+	patternLevels := strings.Split(pattern, "/")
+
+	for i := 0; i < len(patternLevels); i++ {
+		// If we've run out of topic levels but pattern still has more
+		if i >= len(topicLevels) {
+			// Only valid if current pattern is '#' AND it's the last part
+			// '#' can match zero or more topic levels, so it can match "nothing"
+			return patternLevels[i] == "#" && i == len(patternLevels)-1
+		}
+
+		switch patternLevels[i] {
+		case "#":
+			// '#' must be last in pattern; if so, it matches all remaining topic levels
+			return i == len(patternLevels)-1
+
+		case "+":
+			// '+' matches exactly one topic level, so we just continue to next
+			continue
+
+		default:
+			// If it's not a wildcard, it must match the topic level exactly
+			if patternLevels[i] != topicLevels[i] {
+				return false
+			}
+		}
+	}
+
+	// After processing pattern, topic must not have any extra levels
+	return len(topicLevels) == len(patternLevels)
+}
+
+func (l LDAP) CheckAcl(username, topic, clientid string, acc int32) (bool, error) {
+
+	attributes := []string{}
+
+	if l.AclTopicPatternAttribute != "" {
+		attributes = append(attributes, l.AclTopicPatternAttribute)
+	}
+
+	if l.AclAccAttribute != "" {
+		attributes = append(attributes, l.AclAccAttribute)
+	}
+
+	//If there are no acl attributes defined, assume all privileges for all users.
+	if len(attributes) == 0 {
+		return true, nil
+	}
+
+	//If there is no groupBaseDN, return false.
+	if l.GroupBaseDN == "" {
+		log.Debugf("ldap_group_base_dn not set, cannot check ACL")
```

Comment:
Here is an example, but the comment is in general: if something's an error, I'd log it as such, not at the `debug` level.

Reply:
Good point!